### DISCUSSION TITLE
What's badges should we have in Exercism

### DISCUSSION THREAD

**Discussion by iHiD** [exercism/v3#1719] [General] [answered]
In V3 we're introducing badges - things that you can get/unlock for achieving certain things. They'll appear on your profile to show off things you've done, and you can choose your favourite badge to appear next to your name around the site. Other than that they're entirely meaningless and just for fun. We'll probably show how common/rare they are etc.

What badges would you like to see? Some simple examples to get us started:

- Completed Ruby Concepts (eg "Ruby Learner")
- Completed Ruby Track (eg "Ruby Master")
- Mentored 100 people (eg "Mentoring Master")
- Mentored 1000 people (eg "Mentoring Royalty")
- Track Maintainer
- V1 member
- V2 member

Open to all and any ideas :)

**Comment by SaschaMann** [comment]
- Solve at least one exercise a day X days in a row
- Submit a {docs|exercise creation|testsuite update} PR
- Review more than X PRs
- Open more than X issues
- Submit more than X iterations for one exercise
- Find a bug in an exercise and report it
- Create an exercise from scratch
- Livestream solving an exercise

In video games I tend to find completely unexpected and arbitrary achievements quite amusing. Take these suggestions with more than just a grain of salt:

- Counterspy: Change a privacy preference
- Manage to submit a PR that passes all checks on first try
- Submit a solution after being inactive for X days, X > 6 months
- Night Owl: Submit a solution after midnight (but before ~5-6am)
- Find a bug in the programming language/implementation while solving an exercise and report it
- Code golf: Reduce the length of a solution by more than X% relative to the first iteration

**Reply by BethanyG** [reply]
OMG.  LOVE the arbitrary ones!  YES to those!

Some more:

*  Managed to collect more than X number of emoji on a post in Slack
*  More than two lines of iterations on an exercise (so more than 10, I think?)

**Reply by BethanyG** [reply]
* Longest (_friendly_) debate on a submitted PR

**Reply by SaschaMann** [reply]
- Suggest a badge that is added to the website

**Reply by ignlg** [reply]
Arbitrary ones are a must. In my personal experience, they are the most engaging ones. Some ideas:

- Early Bird: Submit a solution before breakfast (~6am, local time).
- Coffee Rush: Submit N solutions in a single day.

I tend to love those that combine factors / actions / events:

- Hello World: Start a track, submit a solution and be approved in less than T time.
- Back In Town: Login after T time without activity and earn a 5 stars rating the very same day.
- Starting Strong: Submit a solution on January 1st.

**Reply by iHiD** [reply]
I normally have breakfast at about 11am... :D

**Comment by JesseSingleton** [comment]
* On the mentoring metrics I think something like 100, 250, 500, 1000 perhaps? I believe you mentioned that some people had almost hit 10,000??
* Hackathon badges if there are more planned!
* It's a bit boring but time based - 1 year, 2 year etc.
* Streaks - for both mentors and users.

Side note - V1 member should definitely be the old devil logo 😈

**Reply by kotp** [reply]
Don't forget:

- First time mentoring an exercise
  - Got your toes wet!
- First time having one of your exercises mentored through completion.
  - Was that so bad?
- Appreciated, and it feels so good!
  - First 5 star rating from student for mentored exercise
- Bowing out
  - First time you leave a conversation

**Comment by BethanyG** [comment]
* Mentored first person ("Mentoring Minion")
* First Testimonial for Mentoring ("Praiseworthy")
* First 10 Testimonials for Mentoring ("Influencer")
* Mentored first 25 ("Mentoring Aficionado")
* Mentored first 100 ("Mentoring Expert")
* Mentored first 200 ("Mentoring Artisan")
* Mentored first 1000 (_don't like Royalty, but Deity is problematic too_)
* First Contribution to Repo

**Reply by BethanyG** [reply]
*  Tag some exercises as "easter egg" exercises (_at random, or of a certain difficulty level, or....??_) - when a student completes these, they get special 1-time easter egg badges.

*  Tag some student solutions as "easter egg" solutions - when a mentor completes mentoring these, they get a special 1-time easter egg badges.

*  Founders day | V1 day | V2 day | V3 day badges - if you're logged in and mentoring or solving during these, you get a special badge.

*  Contributed to an Analyzer ("Analytical")
*  Main author on an Analyzer
*  Contributed to a test runner ("Suspicious Mind")
*  Main author on a test runner
*  Contributed to a Representer ("Representative")
*  Main author on a Representer

**Comment by neenjaw** [comment]
- X practice exercises complete
- X practice exercises mentor approved
- Made a friend (preferred by a mentor / preferred by a submitter)
- polyglot awards (attempted x languages to some measure)
- some persistence aware (something to reward stick-to-it-ness and growth with revisions)

**Reply by BethanyG** [reply]
> X practice exercises complete 

... _wax on, wax off_

**Comment by angelikatyborska** [comment]
- "[language] Superstar" - one of your solutions for any exercise in a given language received at least X (20? 30? 50?) stars.

- "Neverending Story" - submitted at least 10 iterations to the same exercise, no duplicates.

- "All your base are belong to us" - solved the "All Your Base" exercise.

- "Slow poke" - replied to a comment more than 30 days after the last activity for the given exercise.

**Reply by kotp** [reply]
Some I have seen have hit 20... 10 (iterations) can be fairly common.

**Reply by BethanyG** [reply]
`Die unendliche Geschichte` - and the badge can be the dragon from the movie:

https://www.syfy.com/sites/syfy/files/styles/1140x640_hero/public/neverending-story.jpg

**Comment by neenjaw** [comment]
I think related to @SaschaMann's thoughts about them seeming spontaneous -- It would be an interesting feature to not have them listed, but instead awarded when the condition is triggered.  That way might prevent some of the `gaming behaviours` where people submit hot garbage just to collect all of the badges.  I think dev.to does something like this where the badges are mostly mysterious until you get one.

If then concerned about discoverability, it could be relatively simple to trigger the first few, raising awareness of their existence but mysterious as to what will trigger the next badge.

**Reply by SaschaMann** [reply]
On Steam and other gaming platforms it's common to have both hidden and publicly listed achievements. Something like a "Complete a track" badge could be listed as an achievable badge, while more obscure/creative ones aren't visible until you unlock them.

**Reply by taiyab** [reply]
I envision having ones that even have ???  in the name/description so people have no clue how to get them, but know they exist.

**Reply by dougal** [reply]
Hide the novelties, make the progress or behaviour-encouraging badges visible.

**Reply by SleeplessByte** [reply]
HIDE ALL THE THINGS (but open source the software, lol)

**Reply by iHiD** [reply]
> HIDE ALL THE THINGS (but open source the software, lol)

Yes. So Tai and I discussed this. I was like "someone is going to make a page listing all the badges, so we may as well make that page but just not put it anywhere near the actual list of badges you have".

**Reply by taiyab** [reply]
Or we could hash all the "hidden" ones and have the key in the V3 website repo? Although I think even that is open source...just no contributions...

**Comment by kotp** [comment]
- Mentored 20 per week for 3 weeks running
  - Cheering Section
- Completed 10 exercises in 1 week  (Could be weighted to difficulty level of exercises)
  - On the fast track

**Comment by kotp** [comment]
On Sat, Jun 20, 2020 at 07:54:13AM -0700, BethanyG wrote:
> * Mentored first 1000 (_don't like Royalty, but Deity is problematic too_)
  * Mentored first 1000 ("Like a Boss")

**Comment by iHiD** [comment]
Update: Hating myself for all the code-writing this is going to generate for me :)

**Comment by WeeDom** [comment]
slightly menacing plastercine duck for 1000 mentorships?

**Comment by kotp** [comment]
- Having mentored thrice the quantity of exercises in a week that exceeds the "average required"
  - "Obsession"
- Having mentored thrice the quantity of exercises as above for **n** weeks
  - "Obsessed"

**Reply by joshgoebel** [reply]
More of these please, so I can win them all. :-)

**Reply by iHiD** [reply]
@SleeplessByte suggested to me that everything should be in the form of `2 ** n` as we're programmers :)

**Reply by taiyab** [reply]
@yyyc514 I'm so sorry Josh. I'm contemplating doing a badge for completing every exercise on every track as a legendary tier.

**Comment by taiyab** [comment]
Getting more than 100 replies on a PR.

"Death by a thousand pecks"

**Comment by ErikSchierboom** [comment]
- "Share the love" (mentor a solution)
- Maybe paradigm-specific badges? E.g. "Functional Fanatic" for completing X number of functional languages
- "Exemplar" (have a solution with over 25 stars)
- "Doing it all" (mentor multiple tracks)

And perhaps some harder to quantify badges too:

- "Knowledgeable" (helped tons of students through chat)
- "Reviewer" (helped review PR's)
- "Make it Work" (helped fix bug(s))

**Reply by BethanyG** [reply]
>"Make it Work" (helped fix bug(s))

...._exterminator_

**Comment by SaschaMann** [comment]
Maybe tracks could also define some exercises to unlock a badge. E.g. for completing a specific exercise that teaches a key paradigm of the language.

**Comment by kotp** [comment]
> ...._exterminator_

The Debuginator

**Comment by dougal** [comment]
* "Purgatory" - Write a solution which triggers an execution timeout.
* "First Contact" - Write a comment on someone else's solution.
* "Knuth Reward Cheque" - Report an error in an exercise.
* ~"Bill O'Reilly" - Live stream a solution (with or without bad language).~ _Edit: Fuck this, thanks @SleeplessByte_

**Reply by SleeplessByte** [reply]
(I would personally never want to get the rapist homo hating anti dutch badge, so maybe we use a different name for that last one 🙉 ❤️. I like the idea!!!)

**Reply by dougal** [reply]
Yikes. Wikipedia page is a treat for someone who only knows BOR from that one clip. Suggestion removed.

**Reply by iHiD** [reply]
lol

**Reply by taiyab** [reply]
@dougal I actually came up with the same idea for livestreaming a solution, but I called it "We'll do it live". :D

**Comment by ErikSchierboom** [comment]
- Lackadaisical: solve `bob` in five or more language
- Flawless: first iteration passes all the tests and has no feedback
- Command-liner: submit 10+ exercises through the CLI

**Comment by kotp** [comment]
On Wed, Jun 24, 2020 at 12:35:55PM +0000, Josh Goebel wrote:
> More of these please, so I can win them all. :-)
> 

- Tantalizing
  - Anyone gets it after achieving n badges, unless it is Josh.

  Should keep him motivated to discover why he is missing one.

**Comment by iHiD** [comment, marked as answer]
Thanks everyone. I'm closing in favour of https://github.com/exercism/v3/blob/master/badges.md - follow the instructions there and we can discuss things on individual PRs. I tried to capture everything in this discussion in the bit at the bottom. The only bit I've left out is language-specific stuff or progress things (10 solutions, 50 solutions, etc) as those are probably going to be a different mechanism to badges. If I missed anything else when copying - sorry - please PR! :)

This is fun.
